code/sonalyze/sonarlog/logfile.go

```diff
@@ -0,0 +1,220 @@
+//
+// When a file is appended to, the data to append is added to a list in the LogFile object but no
+// further action is taken.  The file has to be flushed by external action, typically by the Cluster
+// marking the file as dirty and performing periodic flushes of dirty data.
+//
+// A file may cache its data, mostly transparently - in this case, a read operation returns the
+// cached data.  (Not yet implemented.)
+//
+// The files are kept generic through the use of `any`.  We could instead have created a hierarchy
+// of interfaces and/or used templated types but that currently seems like needless complexity.
+//
+// ---- stop reading here ----
+//
+// (Clean this up - not all of it belongs here - it's about the cached state, not yet implemented.)
+//
+// A file is in one of several states:
+//
+// - on disk, no output pending
+// - on disk, output pending
+// - in memory, no output pending
+// - in memory, output pending
+// - locked, state transition ongoing
+//
+// A file can be appended to whether it is on disk or in memory, the output is queued and written
+// asynchronously.
+//
+// When a file is read, it is always brought in from disk to memory.  This does not require flushing
+// its output first.  (Though it would be nice, I guess, to do so.)
+//
+// A file that is on disk with output pending may not actually exist on the disk yet.
+//
+// There may be a soft limit on the number of sample records in memory, this is controlled by a
+// command line switch.  When this switch is present, the soft limit is in effect.  When the switch
+// is not present then no records are held in memory, the file will never be in-memory.
+//
+// When a file has been read and is to be cached, we compute its occupancy O.  If O + the current
+// occupancy of the store, C, exceeds the limit then some older records in the cache have to be
+// purged.  We purge entire files.  Purging is by random-LRU: pick some candidates at random in the
+// cache, then remove the LRU order.  Repeat until we're below the limit again.
+//
+// Purging will not affect records that have been obtained by ongoing operations: They retain their
+// records unchanged.  So the amount of data in memory may for a time exceed the limit.
+
+package sonarlog
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"sync"
+
+	. "sonalyze/common"
+)
+
+const (
+	newline = 10
+)
+
+type fileAttr int
+
+const (
+	fileCacheable fileAttr = 1 << iota
+	fileAppendable
+	fileSonarSamples // `sonar ps` data
+	fileSonarSysinfo // `sonar sysinfo` data
+)
+const (
+	fileContentMask = fileSonarSamples | fileSonarSysinfo
+)
+
+// The components of the Fullname are broken out so as to allow strings to be shared as much as
+// possible.  If we were to catenate them up front we'd be creating a lot of unique strings that
+// would be held permanently.
+//
+// TODO: OPTIMIZEME: These strings could usefully be Ustr, though even with 1M files in memory it
+// probably won't matter very much?
```

Comment:
File a bug for this.  Ustr gives us two things: compact representation and deduplication.  Not obvious that we get deduplication (at least not reliable so) with strings.